File: conf/qmacro/expa.lua
-- Build the merged query
linepat = "^ds:(.-)%.%s+d:(.-)%.%s+t:(.-)%.%s+s:(.-)%.%s+l:(.-)%.%s+v:(.-)%.%s*$"

dataset_aliases = {
	LMSMR = { "lmsmr4x56", "lmsmr4x55", "lmsmr4x57" }
}

Set = {}

function Set:new(o)
	o = o or {}
	setmetatable(o, self)
	self.__index = self
	return o
end

function Set:parse(s)
	set = Set:new()
	for name in s:gmatch("%s*([^,]+)%s*") do
		set[name:gsub("/",",")] = true
	end
	return set
end

function Set:addset(set)
	for k, v in pairs(set) do
		self[k] = true
	end
end

function Set:orexpr()
	res = ""
	first = true
	for k, v in pairs(self) do
		if first then
			res = k
			first = false
		else
			res = res .. " or " .. k
		end
	end
	return res
end

function buildmatcher(s, l, v)
	query = {}
	table.insert(query, "timerange:" .. s:orexpr())
	table.insert(query, "level:" .. l:orexpr())
	table.insert(query, "product:" .. v:orexpr())
	return arki.matcher.new(table.concat(query, "; "))
end


QueryChunk = {}

function QueryChunk:new(o)
	o = o or {}
	setmetatable(o, self)
	self.__index = self
	return o
end

function QueryChunk:init(dsname, d, t)
	self["dsname"] = dsname
	self["date"] = d
	self["time"] = t
	self["queries"] = {}
	-- Parse date
	if d:match("^%s*[Tt]%s*$") then
		d = os.date("%Y-%m-%d")
	elseif d:match("^%s*@%s*[+-]?%s*%d*$") then
		if args == nil then error("If using @, please invoke as --qmacro=\"expa YYYY-MM-DD\"") end
		pm, val = d:match("^%s*@%s*([+-])%s*(%d+)%s*$")
		d = args
		if pm ~= nil then
			rt = os.date("*t")
			rt["year"], rt["month"], rt["day"] = d:match("^(%d%d%d%d)-(%d%d)-(%d%d)$")
			if rt["year"] == nil then error("If using @, please invoke as --qmacro=\"expa YYYY-MM-DD\"") end
			if pm == "+" then
				ts = os.time(rt) + 3600 * 24 * val
			else
				ts = os.time(rt) - 3600 * 24 * val
			end
			d = os.date("%Y-%m-%d", ts)
		end
	elseif d:match("^%d+-%d+-%d+$") then
		-- d is already as we want it
	else
		pm, val = d:match("^%s*[Tt]%s*([+-])%s*(%d+)%s*$")
		if pm == nil then error("Expected 'T+n' or 'T-n'") end
		if pm == "+" then
			ts = os.time() + 3600 * 24 * val
		else
			ts = os.time() - 3600 * 24 * val
		end
		d = os.date("%Y-%m-%d", ts)
	end

	-- Parse time
	h, m = t:match("(%d%d)(%d%d)")
	if h == nil then error("Time expected in form HHMM") end

	self.timestr = d .. string.format(" %02d:%02d:00", h, m)
end

function QueryChunk:addquery(q)
	table.insert(self.queries, q)
end

function QueryChunk:datasets()
	local res = dataset_aliases[self.dsname]
	if res ~= nil then return res end
	return { self.dsname }
end

-- Feed all information we have into a GridQuery
function QueryChunk:setupgq(gq)
	-- Add queries
	for _, q in ipairs(self.queries) do
		gq:add(q)
	end

	-- Add time
	gq:addtime(self.timestr)

	-- Consolidate in the end
	gq:consolidate()

	if debug then
		io.stderr:write("Master query: ", tostring(gq:mergedquery()), "\n")
		io.stderr:write("Result space:\n")
		io.stderr:write(gq:dump(), "\n")
	end
end

function QueryChunk:checkData(cons)
	for idx, dsname in ipairs(self:datasets()) do
		if verbose then io.stderr:write("Trying query on ", dsname, "\n") end

		local ds = qmacro:dataset(dsname)
		if ds == nil then error("Cannot access dataset " .. dsname) end
		local gq = arki.gridquery.new(ds)

		-- Prepare the GridQuery
		self:setupgq(gq)

		if gq:expecteditems() == 0 then
			if verbose then
                                io.stderr:write("No results expected from ", dsname, "\n")
                        end
		else
			-- Build the merged query
			query = gq:mergedquery()
			-- print (query)
			-- print (query:expanded())
		
			-- Query dataset storing results
			self.mds = {}

			ds:queryData({matcher=query}, function(md)
				if gq:checkandmark(md) then
					table.insert(self.mds, md:copy())
				end
				return true
			end)

			
			if #self.mds == 0 then
				if verbose then
					io.stderr:write("No results from ", dsname, "\n")
				end
			elseif gq:satisfied() then
				if verbose then io.stderr:write("Satisfied query on ", dsname, "\n") end
				return
			elseif verbose then
				io.stderr:write("Unsatisfied query on ", dsname, "\n")
				io.stderr:write("Master query: ", tostring(query), "\n")
				io.stderr:write("Result space:\n")
				io.stderr:write(gq:dump(), "\n")
			end
		end
	end

	error("Query cannot be satisfied")
end

function QueryChunk:outputData(cons)
	for idx, md in ipairs(self.mds) do
		cons(md)
	end
end

function QueryChunk:querySummary(sum)
	for idx, dsname in ipairs(self:datasets()) do
		if verbose then io.stderr:write("Getting summary from ", dsname, "\n") end

		local ds = qmacro:dataset(dsname)
		if ds == nil then error("Cannot access dataset " .. dsname) end
		local gq = arki.gridquery.new(ds)

		-- Prepare the GridQuery
		self:setupgq(gq)

		-- Build the merged query
		query = gq:mergedquery()

		-- Add the summary to sum
		ds:querySummary(query, sum)
	end
end


-- Cache of summary information
chunks = {}

-- Parse input
for line in query:gmatch("[^\r\n]+") do
	ds, d, t, s, l, v = line:match(linepat)
	if ds ~= nil then
		s = Set:parse(s)
		l = Set:parse(l)
		v = Set:parse(v)
		q = buildmatcher(s, l, v)

		dstag = ds .. "/" .. d .. "/" .. t
		info = chunks[dstag]
		if chunks[dstag] == nil then
			info = QueryChunk:new()
			info:init(ds, d, t)
			chunks[dstag] = info
		end
		-- print("ADD", name, q)
		info:addquery(q)
	else
		error("line not parsed: " .. line)
	end
end

function queryData(q, cons)
	for _, info in pairs(chunks) do
		info:checkData(cons)
	end
	for _, info in pairs(chunks) do
		info:outputData(cons)
	end
end

function querySummary(q, sum)
	for _, info in pairs(chunks) do
		info:querySummary(sum)
	end
end

File: conf/qmacro/expa.py
# python 3.7+ from __future__ import annotations
import re
import datetime


DATASET_ALIASES = {
    "LMSMR": ["lmsmr4x52", "lmsmr4x55", "lmsmr4x53"],
    "cosmo_5M_ita_any": ["cosmo_5M_ita"],
}


class Row:
    def __init__(self, lineno, line, reftime, dsname, matcher):
        self.lineno = lineno
        self.line = line
        self.reftime = reftime
        self.dsname = dsname
        self.matcher = matcher


class Querymacro:
    re_line = re.compile(r"""
        ^
        ds:(?P<ds>[^.]+)\.\s+  # Dataset
        d:(?P<d>[^.]+)\.\s+    # Date
        t:(?P<t>[^.]+)\.\s+    # Time
        s:(?P<s>[^.]+)\.\s+    # Timerange
        l:(?P<l>[^.]+)\.\s+    # Level
        v:(?P<v>[^.]+)\.\s*    # Product
        $""", re.X)

    re_date_today = re.compile(r"^[Tt]$")
    re_date_today_relative = re.compile(r"^[Tt]\s*(?:(?P<dir>[+-])\s*(?P<days>\d+))?$")
    re_date_relative = re.compile(r"^@\s*(?:(?P<dir>[+-])\s*(?P<days>\d+))?$")
    re_date_iso = re.compile(r"^\d{4}-\d{2}-\d{2}$")
    re_time = re.compile(r"^(?P<h>\d{2})(?P<m>\d{2})$")

    def __init__(self, session, macro_args, query):
        self.session = session

        # The argument, if provided, is a date used to expand @ in date expressions
        macro_args = macro_args.strip()
        if macro_args:
            self.date_ref = datetime.datetime.strptime(macro_args, "%Y-%m-%d").date()
        else:
            self.date_ref = None

        # Rows to query
        self.rows = []

        # Parse query
        for idx, line in enumerate(query.splitlines(), start=1):
            mo = self.re_line.match(line)
            if not mo:
                raise RuntimeError("query:{}: line not parsed: {}".format(idx, repr(line)))

            dsname = mo.group("ds")
            reftime = datetime.datetime.combine(
                    self._to_date(mo.group("d")),
                    self._to_time(mo.group("t")))

            matcher = self.session.matcher(
                "reftime: ={reftime:%Y-%m-%dT%H:%M:%S}Z; "
                "timerange: {trange}; level: {level}; product: {product}".format(
                    reftime=reftime,
                    trange=self._to_matcher(mo.group("s")),
                    level=self._to_matcher(mo.group("l")),
                    product=self._to_matcher(mo.group("v"))
                )
            )

            self.rows.append(Row(idx, line, reftime, dsname, matcher))

        # Sort rows by reftime
        self.rows.sort(key=lambda x: x.reftime)

        # Cache open readers for reuse
        self.readers = {}

    def get_reader(self, name):
        res = self.readers.get(name)
        if res is None:
            res = self.session.dataset_reader(name=name)
            self.readers[name] = res
        return res

    def row_to_md(self, row, with_data):
        datasets = DATASET_ALIASES.get(row.dsname)
        if datasets is None:
            datasets = [row.dsname]
        for ds in datasets:
            reader = self.get_reader(ds)
            mds = reader.query_data(matcher=row.matcher, with_data=with_data)
            if not mds:
                raise RuntimeError("row {}:{} did not produce any results".format(row.lineno, repr(row.line)))
            if len(mds) > 1:
                raise RuntimeError("row {}:{} produced {} results instead of one".format(
                    row.lineno, repr(row.line), len(mds)))
            return mds[0]

    def query_data(self, matcher=None, with_data=False, on_metadata=None):
        # Note: matcher and sort are currently ignored
        # FIXME: change the api to remove them?

        res = []
        for row in self.rows:
            res.append(self.row_to_md(row, with_data))

        if on_metadata:
            for md in res:
                on_metadata(md)
        else:
            return res

    def _to_matcher(self, s):
        return s.replace("/", ",")

    def _to_date(self, d):
        d = d.strip()

        # Today
        mo = self.re_date_today.match(d)
        if mo:
            return datetime.date.today()

        # Relative to today
        mo = self.re_date_today_relative.match(d)
        if mo:
            if mo.group("dir") is None:
                return datetime.date.today()
            elif mo.group("dir") == "+":
                return datetime.date.today() + datetime.timedelta(days=int(mo.group("days")))
            else:
                return datetime.date.today() - datetime.timedelta(days=int(mo.group("days")))

        # Relative to argument
        mo = self.re_date_relative.match(d)
        if mo:
            if not self.date_ref:
                raise RuntimeError("if using @, please invoke as --qmacro=\"expa YYYY-MM-DD\"")

            if mo.group("dir") is None:
                return self.date_ref
            if mo.group("dir") == "+":
                return self.date_ref + datetime.timedelta(days=int(mo.group("days")))
            else:
                return self.date_ref - datetime.timedelta(days=int(mo.group("days")))

        # YYYY-MM-DD
        mo = self.re_date_iso.match(d)
        if mo:
            return datetime.datetime.strptime(d, "%Y-%m-%d").date()

        raise RuntimeError("date expression {} not recognised".format(repr(d)))

    def _to_time(self, t):
        mo = self.re_time.match(t)
        if not mo:
            raise RuntimeError("time {} is not in form HHMM".format(repr(t)))

        return datetime.time(int(mo.group("h")), int(mo.group("m")))
